tests/testsuite/safe_directories.rs

```diff
@@ -0,0 +1,536 @@
+        .masquerade_as_nightly_cargo()
+        .with_stderr(
+            "\
+error: could not load Cargo configuration
+
+Caused by:
+  safe.directories may only be configured from Cargo's home directory
+  Found `safe.directories` in [ROOT]/foo/.cargo/config.toml
+  Cargo's home directory is [ROOT]/home/.cargo
+",
+        )
+        .with_status(101)
+        .run();
+}
+
+#[cargo_test]
+fn home_config_not_checked() {
+    // home config ownership doesn't matter.
+    let p = project().file("src/lib.rs", "").build();
+    let cargo_home = paths::home().join(".cargo");
+    cargo_home.mkdir_p();
+    let home_config = cargo_home.join("config.toml");
+    fs::write(&home_config, "build.jobs = 1").unwrap();
+    p.cargo("check")
+        .env("CARGO_UNSTABLE_SAFE_DIRECTORIES", "true")
+        .env("__CARGO_TEST_OWNERSHIP", &home_config)
+        .masquerade_as_nightly_cargo()
+        .with_stderr(
+            "\
+[CHECKING] foo [..]
+[FINISHED] dev [..]
+",
+        )
+        .run();
+}
+
+#[cargo_test]
+fn environment_variables() {
+    // Check that environment variables are supported.
+    let p = sample_config_project();
+    let cfg = p.root().join(".cargo/config.toml");
+    p.cargo("check")
+        .env("CARGO_UNSTABLE_SAFE_DIRECTORIES", "true")
+        .env("__CARGO_TEST_OWNERSHIP", &cfg)
+        .masquerade_as_nightly_cargo()
+        .with_stderr_contains("[..]`[ROOT]/foo/.cargo/config.toml` is owned[..]")
+        .with_status(101)
+        .run();
+    p.cargo("check")
+        .env("CARGO_UNSTABLE_SAFE_DIRECTORIES", "true")
+        .env("__CARGO_TEST_OWNERSHIP", &cfg)
+        .env("CARGO_SAFE_DIRECTORIES", p.root())
+        .masquerade_as_nightly_cargo()
+        .with_stderr(
+            "\
+[CHECKING] foo [..]
+[FINISHED] dev [..]
+",
+        )
+        .run();
+    p.cargo("check")
+        .env("CARGO_UNSTABLE_SAFE_DIRECTORIES", "true")
+        .env("__CARGO_TEST_OWNERSHIP", &cfg)
+        .env("RUSTUP_SAFE_DIRECTORIES", p.root())
+        .masquerade_as_nightly_cargo()
+        .with_stderr("[FINISHED] dev [..]")
+        .run();
+}
+
+#[cargo_test]
+fn not_allowed_on_cli() {
+    // safe.directories is not supported on the CLI
+    let p = sample_config_project();
+    p.cargo("check -Zunstable-options --config")
+        .env("CARGO_UNSTABLE_SAFE_DIRECTORIES", "true")
+        .arg("safe.directories=['foo']")
+        .masquerade_as_nightly_cargo()
+        .with_stderr("error: safe.directories cannot be set via the CLI")
+        .with_status(101)
+        .run();
+}
+
+#[cargo_test]
+fn cli_path_is_allowed() {
+    // Allow loading config files owned by another user passed on the path.
+    // This is based on the idea that the user has explicitly listed the path,
+    // and that is regarded as intentional trust.
+    let p = project()
+        .file("src/lib.rs", "")
+        .file("myconfig.toml", "build.jobs = 1")
+        .build();
+    p.cargo("check --config myconfig.toml -Zunstable-options")
```

Comment:
`--config`-cl is about to stabilize in 1.63.0

(and ditto throughout)

```suggestion
    p.cargo("check --config myconfig.toml")
```